modules/bash-commons/src/array.sh

```diff
@@ -81,4 +81,48 @@ function array_prepend {
 
   updated_ary=( "${ary[@]/#/$prefix}" )
   echo ${updated_ary[*]}
-}
\ No newline at end of file
+}
+
+# Applies a functional reduce to an array of arguments. Supports mathmatical expressions only (see man expr)
+# The first argument should be a single quoted string representing a statment to evaluate values with. It 
+# expects the following variables, expanded as:
+#   $1 - the reducer, starting at the first value in the array, the resulting value of the entire expression 
+#        becomes this value on the next call
+#   $2 - the next item in the array
+#
+# Synopsis:
+#
+# array_reduce <expression> <elements>...
+#
+# Examples:
+# 
+# array_reduce '$1 * $2' 1 2 3 4 5
+#   Returns: 120
+#
+# Hint: In the expression string, $1 is initialized to the first element in the array, and $2 starts as the second element.
+function array_reduce {
+    local -r expression="$1"
+    shift
+    local reducer="$1"
+    shift
+    local ary=("$@")
+
+    for (( i=0; i<"${#ary[@]}"; i++ )); do
+        echo "${ary[i]}"
+
+        # Expand the expression to be evaluated, by replacing $1 and $2 with the reducer and next array element, respectively
+        # Additionally, make modify '*' to the multiplication token used by expr ('\*') for convience.
+        local expression_expanded="$(printf "$expression" | sed "s/\$1/$reducer/; s/\$2/${ary[i]}/; s/\([^\\]\)\*/\1\\\*/" )"
+        echo $reducer_expanded
+        echo $expression_expanded
+
+        # update the reducer with the result of the next element evaluated by the expression.
+        reducer=$(eval "expr $expression_expanded")
+        if ((  "$?" > "1" )); then
```

Comment:
```suggestion
        if [[  "$?" -ne 0 ]]; then
```

Reply:
Actually, this check won't be necessary in most cases, for if you have `set -e` set, then the script will automatically exit on a non-zero exit code.

Reply:
Makes sense. I've decided to leave it in for now, since `set -e` could still be absent if a user didn't declare it.